### PR TITLE
go1.14 support

### PR DESCRIPTION
**Pull request by cosnicolaou**
This PR provides go 1.14 compatibility. There are two changes:
1. calls to testing.Init() to initialize flags as required for regression tests
2. in celebration of go modules being declared 'production' ready, ensure that vdl can operate in a go module directory structure.

### CODE REVIEW

File: x/ref/cmd/vdl/main.go [hunk 1]
```diff
@@ -668,19 +676,42 @@ func handleErrorOrSkip(prefix string, err error, env *compile.Env) bool {
 
 var errSkip = fmt.Errorf("SKIP")
 
+// Handle the case where go modules are used the directory structure
+// on the local filesystem omits the portion of the package path
+// represented by the module definition in the go.mod file.
```
Comment (razvanm) [inline x/ref/cmd/vdl/main.go]: Are some words missing from this sentence?

File: x/ref/lib/vdl/build/build.go [hunk 2]
```diff
@@ -311,6 +312,68 @@ func SrcDirs(errs *vdlutil.Errors) []string {
 	return append(srcDirs, vdlPathSrcDirs(errs)...)
 }
 
+// HasGoModule returns the value of the module statement in
+// the go.mod file in the directory specified by path, or an empty
+// string otherwise.
```
Comment (razvanm) [inline x/ref/lib/vdl/build/build.go]: Reflow?

File: x/ref/lib/vdl/build/build.go [hunk 3]
```diff
@@ -311,6 +312,68 @@ func SrcDirs(errs *vdlutil.Errors) []string {
...
+
+// PackagePathSplit returns the longest common suffix in dir and path,
+// the prefix in dir after the common suffix is removed and the portion of the
+// of the original path that is not part of the common prefix. This is useful
+// for working with the directory structure commonly used by go modules.
+// For example:
+// PackagePathSplit("a/b/c", "/b/c") yields "a", "", "b/c"
+// PackagePathSplit("a/b/c", "m/b/c") yields "a", "m", "b/c"
+// It is intended to work on Windows, as in:
+// PackagePathSplit(`a\b\c`, "m/b/c") yields "a", "m", "b/c"
```
Comment (razvanm) [inline x/ref/lib/vdl/build/build.go]: Does this show up the way you want in godoc?

File: x/ref/lib/vdl/build/build.go [hunk 4]
```diff
@@ -596,20 +668,46 @@ func (ds *depSorter) resolveWildcardPath(isDirPath bool, prefix, suffix string)
 				if strings.HasPrefix(pkgPath, vdlrootImportPrefix) {
 					return filepath.SkipDir
 				}
+
+				// Special case to handle go modules, where the current
+				// directory is wuthin a go module and hence the matching
```
Comment (razvanm) [inline x/ref/lib/vdl/build/build.go]: s/wuthin/within/

File: x/ref/lib/vdl/build/build.go [hunk 5]
```diff
@@ -311,6 +312,68 @@ func SrcDirs(errs *vdlutil.Errors) []string {
...
+// the prefix in dir after the common suffix is removed and the portion of the
+// of the original path that is not part of the common prefix. This is useful
+// for working with the directory structure commonly used by go modules.
+// For example:
+// PackagePathSplit("a/b/c", "/b/c") yields "a", "", "b/c"
+// PackagePathSplit("a/b/c", "m/b/c") yields "a", "m", "b/c"
+// It is intended to work on Windows, as in:
+// PackagePathSplit(`a\b\c`, "m/b/c") yields "a", "m", "b/c"
+func PackagePathSplit(dir, pkgPath string) (prefix string, body string, suffix string) {
+	// Normalize dir path to use / instead of possibly using
```
Comment (razvanm) [inline x/ref/lib/vdl/build/build.go]: '.' at the end. :-)

File: x/ref/lib/vdl/build/build.go [hunk 6]
```diff
@@ -311,6 +312,70 @@ func SrcDirs(errs *vdlutil.Errors) []string {
 	return append(srcDirs, vdlPathSrcDirs(errs)...)
 }
 
+// HasGoModule returns the value of the module statement in the go.mod file in
+// the directory specified by path, or an empty string otherwise.
+func HasGoModule(path string) string {
+	buf, err := ioutil.ReadFile(filepath.Join(path, "go.mod"))
+	path = filepath.Dir(path)
+	if err != nil && os.IsNotExist(err) {
```
Comment (siddharthab) [inline x/ref/lib/vdl/build/build.go]: What if the error is of some other kind?

File: x/ref/lib/vdl/build/build.go [hunk 7]
```diff
@@ -311,6 +312,70 @@ func SrcDirs(errs *vdlutil.Errors) []string {
 	return append(srcDirs, vdlPathSrcDirs(errs)...)
 }
 
+// HasGoModule returns the value of the module statement in the go.mod file in
+// the directory specified by path, or an empty string otherwise.
+func HasGoModule(path string) string {
```
Comment (siddharthab) [inline x/ref/lib/vdl/build/build.go]: nit: The name of this function is a bit confusion. Maybe consider `GoModuleName`?

File: x/ref/lib/vdl/build/build.go [hunk 8]
```diff
@@ -596,20 +670,46 @@ func (ds *depSorter) resolveWildcardPath(isDirPath bool, prefix, suffix string)
 				if strings.HasPrefix(pkgPath, vdlrootImportPrefix) {
 					return filepath.SkipDir
 				}
+
+				// Special case to handle go modules, where the current
+				// directory is within a go module and hence the matching
+				// for the vdlroot directory must take this into account.
+				for goModRoot, goModPrefix := range ds.goModules {
+					// Test for dirPath being within s go module.
```
Comment (siddharthab) [inline x/ref/lib/vdl/build/build.go]: s/s/a/

File: x/ref/lib/vdl/build/build.go (hunk 7 above)
Comment (cosnicolaou) [reply x/ref/lib/vdl/build/build.go]: agreed!

File: x/ref/lib/vdl/build/build.go (hunk 6 above)
Comment (cosnicolaou) [reply x/ref/lib/vdl/build/build.go]: fixed to return an error.

**Comment by cosnicolaou** [comment]
On Sun, Mar 1, 2020 at 11:46 AM Siddhartha Bagaria <notifications@github.com>
wrote:

> *@siddharthab* approved this pull request.
>
> Thank you for this change!
>
> I am not very sure of how the current code changes wire into the larger
> system, but they look fine by themselves. Let me know if you want me to do
> a more detailed review.
>

thanks for the review and there's no need for you to dig deeper. The main
benefit of this PR is that it avoids the need to lay out the vanadium code
the awkward ../src/... manner.

Cheers, Cos.

File: x/ref/cmd/vdl/main.go [hunk 9]
```diff
@@ -101,9 +101,12 @@ Import path elements and file names are not allowed to begin with "." or "_";
...
 
-Note that whereas GOPATH requires *.go source files and packages to appear
-under a "src" directory, VDLPATH requires *.vdl source files and packages to
-appear directly under the VDLPATH directories.
+
+
+VDLPATH requires *.vdl source files and packages to appear directly under the
+VDLPATH directories. Note that when go modules are used VDLPATH should point
+to the location of the go.mod file. Also note that whereas GOPATH requires
+*.go source files and packages to appear under a "src" directory,
```
Comment (tatatodd) [inline x/ref/cmd/vdl/main.go]: Minor: we're missing the last part of the last sentence.

File: x/ref/runtime/internal/rpc/testutil_test.go [hunk 10]
```diff
@@ -7,7 +7,7 @@ package rpc
 import (
 	"time"
 
-	"v.io/v23"
+	v23 "v.io/v23"
```
Comment (tatatodd) [inline x/ref/runtime/internal/rpc/testutil_test.go]: Are these local package names necessary?  Shouldn't the name of the package be `v23` anyways?

Comment (cosnicolaou) [reply x/ref/runtime/internal/rpc/testutil_test.go]: this is just goimports fixing the ambiguity between go module version #s and v23 which is a valid version number!

**Pull request merged**